executor/load_data.go

```diff
@@ -142,6 +143,232 @@ func NewLoadDataWorker(
 		setNonRestrictiveFlags(userSctx.GetSessionVars().StmtCtx)
 	}
 
+	loadDataWorker := &LoadDataWorker{
+		UserSctx:   userSctx,
+		table:      tbl,
+		controller: controller,
+		planInfo: planInfo{
+			ID:          plan.ID(),
+			Columns:     plan.Columns,
+			GenColExprs: plan.GenCols.Exprs,
+		},
+		progress: asyncloaddata.NewProgress(),
+	}
+	return loadDataWorker, nil
+}
+
+func (e *LoadDataWorker) loadRemote(ctx context.Context) (int64, error) {
+	if err2 := e.controller.InitDataFiles(ctx); err2 != nil {
+		return 0, err2
+	}
+	return e.load(ctx, nil)
+}
+
+// LoadLocal reads from client connection and do load data job.
+func (e *LoadDataWorker) LoadLocal(ctx context.Context, r io.ReadCloser) error {
+	_, err := e.load(ctx, r)
+	return err
+}
+
+func (e *LoadDataWorker) load(ctx context.Context, r io.ReadCloser) (int64, error) {
+	var (
+		job *asyncloaddata.Job
+		err error
+	)
+
+	s, err := CreateSession(e.UserSctx)
+	if err != nil {
+		return 0, err
+	}
+	defer func() {
+		// if the job is detached, we should not close the session. let go-routine to do it.
+		if !e.controller.Detached {
+			CloseSession(s)
+		}
+	}()
+
+	sqlExec := s.(sqlexec.SQLExecutor)
+
+	job, err = asyncloaddata.CreateLoadDataJob(
+		ctx,
+		sqlExec,
+		e.GetInfilePath(),
+		e.controller.DBName,
+		e.table.Meta().Name.O,
+		e.controller.ImportMode,
+		e.UserSctx.GetSessionVars().User.String(),
+	)
+	if err != nil {
+		return 0, err
```

Comment:
in these exit branch with true `e.controller.Detached` we will leak the session

Reply:
https://github.com/pingcap/tidb/pull/42883/commits/d704f39d044d3e91ef10bd4684354e118caec594